packages/core-data/src/fetch/fetch-link-suggestions.ts

```diff
@@ -0,0 +1,356 @@
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import { addQueryArgs } from '@wordpress/url';
+import { decodeEntities } from '@wordpress/html-entities';
+import { __ } from '@wordpress/i18n';
+
+export type SearchOptions = {
+	/**
+	 * Displays initial search suggestions, when true.
+	 */
+	isInitialSuggestions?: boolean;
+	/**
+	 * Search options for initial suggestions.
+	 */
+	initialSuggestionsSearchOptions?: Omit<
+		SearchOptions,
+		'isInitialSuggestions' | 'initialSuggestionsSearchOptions'
+	>;
+	/**
+	 * Filters by search type.
+	 */
+	type?: 'attachment' | 'post' | 'term' | 'post-format';
+	/**
+	 * Slug of the post-type or taxonomy.
+	 */
+	subtype?: string;
+	/**
+	 * Which page of results to return.
+	 */
+	page?: number;
+	/**
+	 * Search results per page.
+	 */
+	perPage?: number;
+};
+
+export type EditorSettings = {
+	/**
+	 * Disables post formats, when true.
+	 */
+	disablePostFormats?: boolean;
+};
+
+type SearchAPIResult = {
+	id: number;
+	title: string;
+	url: string;
+	type: string;
+	subtype: string;
+};
+
+type MediaAPIResult = {
+	id: number;
+	title: { rendered: string };
+	source_url: string;
+	type: string;
+};
+
+export type SearchResult = {
+	/**
+	 * Post or term id.
+	 */
+	id: number;
+	/**
+	 * Link url.
+	 */
+	url: string;
+	/**
+	 * Title of the link.
+	 */
+	title: string;
+	/**
+	 * The taxonomy or post type slug or type URL.
+	 */
+	type: string;
+	/**
+	 * Link kind of post-type or taxonomy
+	 */
+	kind?: string;
+};
+
+/**
+ * Fetches link suggestions from the WordPress API.
+ *
+ * WordPress does not support searching multiple tables at once, e.g. posts and terms, so we
+ * perform multiple queries at the same time and then merge the results together.
+ *
+ * @param search
+ * @param searchOptions
+ * @param editorSettings
+ *
+ * @example
+ * ```js
+ * import { fetchLinkSuggestions } from '@wordpress/core-data';
+ *
+ * //...
+ *
+ * export function initialize( id, settings ) {
+ *
+ * settings.__experimentalFetchLinkSuggestions = (
+ *     search,
+ *     searchOptions
+ * ) => fetchLinkSuggestions( search, searchOptions, settings );
+ * ```
+ */
+export default async function fetchLinkSuggestions(
+	search: string,
+	searchOptions: SearchOptions = {},
+	editorSettings: EditorSettings = {}
+): Promise< SearchResult[] > {
+	if (
+		searchOptions.isInitialSuggestions &&
+		searchOptions.initialSuggestionsSearchOptions
+	) {
+		searchOptions = {
+			...searchOptions,
+			...searchOptions.initialSuggestionsSearchOptions,
+		};
```

Comment:
Removed the mutation 👍 

Thinking about it though, `isInitialSuggestions` is a pretty strange API design. It should be the callers responsibility... I'll look at cleaning this up since we're stabilising.

Reply:
Later's problem https://github.com/WordPress/gutenberg/pull/62397#issuecomment-2167265679